alica_engine/src/engine/AlicaTimer.cpp

```diff
@@ -0,0 +1,78 @@
+#include "engine/AlicaTimer.h"
+
+#include <thread>
+
+namespace alica
+{
+
+class AlicaSystemTimer::AlicaSystemTimerImpl
+{
+public:
+    AlicaSystemTimerImpl(TimerCb&& userCb, alica::AlicaTime period)
+            : _userCb(std::move(userCb))
+            , _period(period.inMilliseconds())
+            , _isActive(false)
+    {
+    }
+
+    ~AlicaSystemTimerImpl()
+    {
+        if (_isActive) {
+            stop();
+        }
+    }
+
+    void start()
+    {
+        if (_isActive) {
+            stop();
+        }
+        _isActive = true;
+        _thread = std::thread([this]() {
+            using namespace std::chrono;
+            while (_isActive) {
+                auto hasRepeatedInternalLoop = false;
+                auto startOuterLoop = system_clock::now();
+                _userCb();
+                int64_t sleep_duration = _period;
+                system_clock::time_point startInnerLoop;
+                // This second while is a trick to loop more than once, in 500ms intervals,
+                // while we wait for sleep_duration milliseconds. This allows us to exit this
+                // loop earlier without having to wait for the full duration when we Ctrl+C and kill this process.
+                while ((sleep_duration > 0) && _isActive) {
+                    auto loop_duration = duration_cast<milliseconds>((system_clock::now() - (hasRepeatedInternalLoop ? startInnerLoop : startOuterLoop)));
```

Comment:
If we just initialize startInnerLoop to startOuterloop the ?: operator isn't needed

Reply:
nit:  I think overall logic here would be a bit clearer if we just calculated the loop end time from the start and compared with that, rather than updating the "start" time

Reply:
Addressed the first comment. About the second,

sorry if I didn't understand correctly, but I think it's not possible. Internally after the first pass where I have the outer loop exec time, I will have the sleep of (500ms - exec time). When I go to the second loop, if I calculate the entire thing I would have something around 500ms + 30ms (exec time of inner loop), and I need to sleep for (500ms - loop_duration (which is 30ms)), so I would need to subtract 500ms from loop_duration, and then 1000ms for the next loop, and 1500ms and so on. I understand what you want can be done, but it would be as complicated as it is right now or even worse because it would be less natural (I would have to decrease loop_duration by (500ms * innerLoopIterations) when sleeping by (500ms - loop_duration)).

unless I get rid of this inner loop exec time and consider it 500ms every time, then it could be done easily, but it would be less correct.

Reply:
Not really sure I understand your comment.  The logic I was thinking was something like
```
f(sleep_time):
    auto now = Time::now();
    auto end_time = now + sleep_time
    while(now < end_time && _isRunning) {
       if(now + 500ms < end_time) {
           sleep(500ms);
      } else {
          sleep(now - end_time);
     }
}
```

Reply:
Changed